internal/pkg/runtime/engines/singularity/process_linux.go

```diff
@@ -35,6 +38,24 @@ func (engine *EngineOperations) checkExec() error {
 		shell = "/bin/sh"
 	}
 
+	// Make sure the shell exists.
+	if _, err := os.Stat(shell); os.IsNotExist(err) {
+		return fmt.Errorf("shell %s doesn't exist in container", shell)
+	}
+
+	// Verify architecture of image matches architecture of Go.
```

Comment:
We can't assume image is always SIF, could be a sandbox, a single squashfs/ext3 image. After thinking, architecture check should be after `syscall.Exec` instead and if `err == ENOENT` because we need to be able to run things like `singularity shell -B /usr/bin/qemu-ppc64le-static docker://ppc64le/ubuntu`, and adding architecture check here will always fail for this use case